halo2-base/src/virtual_region/copy_constraints.rs

```diff
@@ -0,0 +1,146 @@
+use std::any::TypeId;
+use std::collections::{BTreeMap, HashMap};
+use std::ops::DerefMut;
+use std::sync::{Arc, Mutex, OnceLock};
+
+use itertools::Itertools;
+use rayon::slice::ParallelSliceMut;
+
+use crate::halo2_proofs::{
+    circuit::{Cell, Region},
+    plonk::{Assigned, Column, Fixed},
+};
+use crate::utils::halo2::{raw_assign_fixed, raw_constrain_equal, Halo2AssignedCell};
+use crate::AssignedValue;
+use crate::{ff::Field, ContextCell};
+
+use super::manager::VirtualRegionManager;
+
+/// Thread-safe shared global manager for all copy constraints.
+pub type SharedCopyConstraintManager<F> = Arc<Mutex<CopyConstraintManager<F>>>;
+
+/// Global manager for all copy constraints. Thread-safe.
+///
+/// This will only be accessed during key generation, not proof generation, so it does not need to be optimized.
+///
+/// Implements [VirtualRegionManager], which should be assigned only after all cells have been assigned
+/// by other managers.
+#[derive(Clone, Default, Debug)]
+pub struct CopyConstraintManager<F: Field + Ord> {
```

Comment:
We should move this out because `CopyConstraintManager ` is more than `VirtualRegionManager` which "is responsible for managing a virtual region and assigning the virtual region to a physical Halo2 region". 

What we actually need here is `ConstantManager`.

Adding equality constraints could be move out from `assign_raw` because it should be always called in the end so we don't need the abstraction.

Reply:
`ConstantManager` is tied together with `constant_equalities` however: you only use constants when you want to enable an equality, in which case you need to tag it and then record that the equality needs to be imposed later. Splitting out `ConstantManager` from this would require extra tagging and duplicate work.